### ISSUE TITLE
How to add (weight, probability and magnitude) to the forward calculation to calculate the gradient？

### ISSUE THREAD

**Issue by hwx10** [VDIGPKU/DADA#14]
I have the following two questions. Is the code in No.1 used to calculate the gradient of magnitude and probability when backward? And the second code is used to calculate the gradient of ops_weights?

1、
    def forward(self, origin_images, trans_images, probability, probability_index, magnitude):
        index = sum( p_i.item()<<i for i, p_i in enumerate(probability_index))
        com_image = 0
        images = origin_images
        adds = 0

        for selection in range(2**len(self.sub_policy)):
            trans_probability = 1
            for i in range(len(self.sub_policy)):
                if selection & (1<<i):
                    trans_probability = trans_probability * probability[i]
                   
                    if selection == index:
                        images = images - magnitude[i]
                        adds = adds + magnitude[i]
                else:
                    trans_probability = trans_probability * ( 1 - probability[i] )
            if selection == index:
                images = images.detach() + adds
                com_image = com_image + trans_probability * images
            else:
                com_image = com_image + trans_probability
        return com_image  

2、
def forward(self, origin_images, trans_images_list, probabilities, probabilities_index, magnitudes, weights, weights_index):
       
        for i, (p, p_i, m, w, op) in enumerate(zip(probabilities, probabilities_index, magnitudes, weights, self._ops)):
            if weights_index.item() == i:
                return sum(w * op(origin_images, trans_images, p, p_i, m))
            else:
                return w

**Comment by latstars** [comment]
The gradient is auto-calculated when backward, since the use of images.detach() + adds.
Two paths exist.
In the first path, the gradient will be backward to com_image, then to images, then to adds, then to probability[i] and magnitude[i].
In the second path, the gradient will be backward to w, weights, op_weights.

**Comment by Trent-tangtao** [comment]
hello, I have some question in search_relax.
> parser.add_argument('--use_parallel', type=bool, default=False, help="use data parallel default False")
The default valse is False, and  in
>  def forward(self, origin_images, probabilities_b, magnitudes, weights_b):
the args are probabilities_b, magnitudes, weights_b, how to calculate the gradient of model.probabilities and model.ops_weights

**Comment by ZengYaoJ** [comment]
@Trent-tangtao did you resolve this problem？  how to calculate the gradient of model.probabilities and model.ops_weights
